Source/WebCore/Modules/permissions/Permissions.cpp

```diff
@@ -77,67 +82,72 @@ static bool isAllowedByFeaturePolicy(const Document& document, PermissionName na
         return;
     }
 
-    if (is<Document>(context) && !isAllowedByFeaturePolicy(downcast<Document>(*context), parameterDescriptor.name)) {
-        context->postTask([parameterDescriptor, promise = WTFMove(promise)](auto& context) mutable {
-            promise.resolve(PermissionStatus::create(context, PermissionState::Denied, parameterDescriptor));
+    auto* origin = context->securityOrigin();
+    auto originData = origin ? origin->data() : SecurityOriginData { };
+
+    if (document) {
+        if (!document->page()) {
+            promise.reject(Exception { InvalidStateError, "The page does not exist"_s });
+            return;
+        }
+
+        if (!isAllowedByFeaturePolicy(*document, permissionDescriptor.name)) {
+            promise.resolve(PermissionStatus::create(*context, PermissionState::Denied, permissionDescriptor));
+            return;
+        }
+
+        PermissionController::shared().query(ClientOrigin { document->topOrigin().data(), WTFMove(originData) }, PermissionDescriptor { permissionDescriptor }, *document->page(), [document = Ref { *document }, permissionDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
+            if (!permissionState)
+                promise.reject(Exception { NotSupportedError, "Permissions::query does not support this API"_s });
+            else
+                promise.resolve(PermissionStatus::create(document, *permissionState, permissionDescriptor));
         });
         return;
     }
 
-    auto* origin = context->securityOrigin();
-    auto originData = origin ? origin->data() : SecurityOriginData { };
+    auto* workerGlobalScope = dynamicDowncast<WorkerGlobalScope>(*context);
+    ASSERT(workerGlobalScope);
 
-    PermissionController::shared().query(ClientOrigin { context->topOrigin().data(), originData }, PermissionDescriptor { parameterDescriptor }, *document->page(), [this, protectedThis = Ref { *this }, parameterDescriptor, promise = WTFMove(promise)](auto permissionState) mutable {
-        auto context = m_navigator ? m_navigator->scriptExecutionContext() : nullptr;
-        if (!context || !context->globalObject())
-            return;
+    auto completionHandler = [originData = WTFMove(originData).isolatedCopy(), permissionDescriptor, contextIdentifier = workerGlobalScope->identifier(), promise = WTFMove(promise)] (auto& context) mutable { 
+        ASSERT(isMainThread());
 
-        context->postTask([parameterDescriptor, promise = WTFMove(promise),  permissionState = WTFMove(permissionState)](auto& context) mutable {
-            if (!permissionState) {
-                promise.reject(Exception { NotSupportedError });
-                return;
-            }
-            promise.resolve(PermissionStatus::create(context, *permissionState, parameterDescriptor));
+        auto& document = downcast<Document>(context);
+        if (!document.page()) {
+            promise.reject(Exception { InvalidStateError, "The page does not exist"_s });
```

Comment:
This is still rejecting the promise on the wrong thread.

Reply:
Fixed!